### ISSUE TITLE
[Feature]: Support elasticsearch data stream

### ISSUE THREAD

**Issue by pengweiqhca** [jaegertracing/jaeger#4708]
### Requirement

https://www.elastic.co/guide/en/elasticsearch/reference/current/data-streams.html

### Problem

I has created a data stream index template, and I tried `--es.use-aliases=true` but not work.

**Comment by jack78901** [comment]
Is this because the client library used for connecting to Elasticsearch is outdated? Namely, I see in the logs of version 1.54 of the collector that the client is olivere/elastic@v6.2.37. This client is from July 27th, 2021. The client is also not being updated anymore, it would appear. Namely, it was last updated to version 7.0.32 on Mar 19th, 2022. This is almost two years ago now. Elastic has its own go library now: https://github.com/elastic/go-elasticsearch.

**Comment by tronda** [comment]
I see that the OpenSearch project has the [same feature-request](https://github.com/opensearch-project/observability/issues/1541).

**Comment by JaredTan95** [comment]
I am currently working on this feature, but I prefer to implement it in jaeger v2

**Comment by yurishkuro** [comment]
@JaredTan95 +2 for v2, but what specifically is the difference? Our storage implementations are currently identical in v1 and v2 (with the exception of configuration)

**Comment by yurishkuro** [comment]
1. that doesn't really address my question - all ES-specific details are hidden behind `es.Factory`, which is used by both Jaeger v1 and v2, so I don't see a difference
2. Data streams also need to be set up once, just like index mappings. We want to do that on collector start-up, with some coordination between multiple collector instances (if creation operation are idempotent then it's less of an issue)

**Comment by Evesy** [comment]
@JaredTan95 Is this still something you're looking at/working on?

We're also keen to look at using data streams for Jaeger. In our case we manage all the templates, lifecycle etc. outside of Jaeger, so our main requirements are to have Jaeger not add the date suffix to the index it writes to, and also to use only `create` operations when indexing (since `index` op type is not allowed for datastreams)

>Data streams also need to be set up once, just like index mappings
You can get by without explicitly creating the datastream up front and just having a matching index template that is configured to use a data stream

**Comment by zzzk1** [comment]
@yurishkuro @Evesy Follow [this guide](https://www.elastic.co/guide/en/elasticsearch/reference/current/tutorial-manage-new-data-stream.html). We have two steps to complete: 1. [Create an index template](https://www.elastic.co/guide/en/elasticsearch/reference/current/tutorial-manage-new-data-stream.html#create-index-template-with-lifecycle) 2. [Create a data stream](https://www.elastic.co/guide/en/elasticsearch/reference/current/tutorial-manage-new-data-stream.html#create-data-stream-with-lifecycle). 
We can utilize existing resources to implement our ideas.
1. To create a template using `es\client.CreateTemplate()`, we need the following components: `mapping`, `setting`, and `index`.
2. A new call data stream client sent an HTTP request: `PUT _data_stream/my-data-stream` to create it.

Do you have any advice? 🤔

**Comment by yurishkuro** [comment]
What sort of advice are you asking for?

**Comment by zzzk1** [comment]
> What sort of advice are you asking for?

follow: 
1. run Jaeger locally, execute the command `/jaeger/cmd/all-in-one/main.go` with Elasticsearch as the backend storage.
2. Sent a GET request to `{{ES_LOCAL_URL}}/jaeger-span-2025-01-15/_search?pretty` and referenced [span-response.json](https://github.com/user-attachments/files/18419301/span-response.json). 

Please take note of the following:
Index Template: 1. jaeger-dependencies 2. jaeger-span 3. jaeger-service Only **jaeger-span-2025-01-15** contains a timestamp field. Should we create a data stream for **jaeger-span** only?

**Comment by yurishkuro** [comment]
Yes we can start with that. Services index is just a cache to avoid scanning the whole database to find distinct service names - perhaps ES can maintain that automatically as some sort of materialized preaggregation.

**Comment by Manik2708** [comment]
@yurishkuro @zzzk1 I have been studying about data streams for a while and I think these points need to be covered:
1. I don't think we need to extend the support for data stream in `es-rollover`, wouldn't it be better to integrate ILM/ISM and data streams directly with jaeger binary and depreceating the mannual rollover. This also refers to: #6283.
2. Using ILM with data streams would be ideal but auto creating policy is challenging, as the policy for backing stream indexes and automatic rollover will look like (as per official docs):
```
{
  "policy": {
    "phases": {
      "hot": {
        "actions": {
          "rollover": {
            "max_primary_shard_size": "50gb"
          }
        }
      },
      "warm": {
        "min_age": "30d",
        "actions": {
          "shrink": {
            "number_of_shards": 1
          },
          "forcemerge": {
            "max_num_segments": 1
          }
        }
      },
      "cold": {
        "min_age": "60d",
        "actions": {
          "searchable_snapshot": {
            "snapshot_repository": "found-snapshots"
          }
        }
      },
      "frozen": {
        "min_age": "90d",
        "actions": {
          "searchable_snapshot": {
            "snapshot_repository": "found-snapshots"
          }
        }
      },
      "delete": {
        "min_age": "735d",
        "actions": {
          "delete": {}
        }
      }
    }
  }
}
```
Now making the policy configurable for the user is going to be a challenge. We can choose to create our own policy struct (because it is not provided by ES-Client) but that will lead to maintainence overhead as keeping it always synchornized with ES updates will be challenging.
3. What part of policy needs to be configurable? I am not in support of creating the whole policy configurable, take an example when using data streams we should not skip rollover.
4. Is this change backward compatible? probably yes! because reading from the data stream is same as reading from index but the real question is when dealing with old data will it work as expected? Users updating to newer version will face two kind of data, one through old index and another through data streams. I am searching for this but currently unclear whether this is an acceptable situation or not.
5. Will data streams be an option or is jaeger going to use only data streams in future? This will answer this question: Should we update the existing index templates to make them compatible with data streams or create new templates so that if user enables data streams they could come in action.

**Comment by Manik2708** [comment]
> [@JaredTan95](https://github.com/JaredTan95) Is this still something you're looking at/working on?
> 
> We're also keen to look at using data streams for Jaeger. In our case we manage all the templates, lifecycle etc. outside of Jaeger, so our main requirements are to have Jaeger not add the date suffix to the index it writes to, and also to use only `create` operations when indexing (since `index` op type is not allowed for datastreams)

I don't think it will break your requirements. Adding a data stream will not change the naming convention of index of jaeger! As far what could I understand from the docs, it will only require adding `@timestamp` in mapping https://www.elastic.co/guide/en/elasticsearch/reference/current/set-up-a-data-stream.html#create-component-templates

**Comment by data-dude** [comment]
Using ILM with datastreams makes sense to me.  People are definitely going to want to configure age values.  Datastreams makes the system more robust and easy to maintain.  I'm not sure why someone wouldn't want to use it for something like traces.   But some people may still want to use the old way, I dunno.

**Comment by yurishkuro** [comment]
@Manik2708 what if we didn't support any automation or customization? We can include a sample JSON for creating a policy, if people need to customize it then can. The only automation that I think will be useful is if the policy was automatically created when Jaeger starts the first time against empty storage, by using that same JSON file. We can provide an option for a user to override that file _as a whole_, if they want (they can also edit the policy via ES JSON API, I assume).

**Comment by Manik2708** [comment]
> [@Manik2708](https://github.com/Manik2708) what if we didn't support any automation or customization? We can include a sample JSON for creating a policy, if people need to customize it they can. The only automation that I think will be useful is if the policy was automatically created when Jaeger starts the first time against empty storage, by using that same JSON file. We can provide an option for a user to override that file _as a whole_, if they want (they can also edit the policy via ES JSON API, I assume).

This sounds good! But I have a question, Where are we gonna keep this json file so that user can edit/override that? I am assuming to keep it in factory just like mappings but then it will become difficult for the user to find that file! 

>  (they can also edit the policy via ES JSON API, I assume).

Yes there is an API

**Comment by yurishkuro** [comment]
> Where are we gonna keep this json file so that user can edit/override that

It will be in this repository and embedded in the binary. The documentation on the website can link to it, while explaining how to apply that policy against the db.

**Comment by zzzk1** [comment]
@yurishkuro @Manik2708 I write a doc about how i am try to implement this feature, please see: [Jaeger ES data stream proposal](https://docs.google.com/document/d/1A4ExzepAs9439myailp_l-yemwuCZXn-M5Wj9OcwaCQ/edit?usp=sharing) thanks.

**Comment by Manik2708** [comment]
> [@yurishkuro](https://github.com/yurishkuro) [@Manik2708](https://github.com/Manik2708) I write a doc about how i am try to implement this feature, please see: [Jaeger ES data stream proposal](https://docs.google.com/document/d/1A4ExzepAs9439myailp_l-yemwuCZXn-M5Wj9OcwaCQ/edit?usp=sharing) thanks.

Things look good to me but some points over it:
1. We use two clients in ES: Olivere and the v8 client. There is no policy type in olivere but it is present in official client. Everywhere olivere client is used except in creating templates for v8 of ES. Even if we find a way still we can't embed the whole policy struct inside the config. Please see https://github.com/jaegertracing/jaeger/issues/4708#issuecomment-2599442335. This seems to be a good approach and later user can even update the policy by using ES APIs.
2. In OS, there is ISM policy instead of ILM policy so we have to take care of that also (although they are vey similar).
3. I have been implementing the ILM policy in jaeger and it is currently in E2E testing phase (it's very long PR (1500 lines) so currently testing it strictly) and the question which is always in my mind is that: Jaeger allows mannual installation of templates so what if user has enabled ILM but is installing templates manually? What if those templates (installed by user) become problem for rollover as it might not include policy name or rollover alias. The only solution which I could think of that is if user enables ILM then mannual installation of templates could not be allowed or we have to update the template according to the use of ILM. I think @yurishkuro can help us in this.
4. Will the policy which you have included in the docs be enough for the datastream? There are many phases and we have to discuss them thoroughly for what can create an effective data stream.
5. I have researched about ILM but not on data-streams currently but we need to ensure that whatever change we are introducing is backward compatible.

**Comment by yurishkuro** [comment]
@zzzk1 please allow comments on the doc.

@Manik2708 
  1. As discussed previously, I do not want to invent yet another configuration language to describe ILM policy. We should only support the official JSON representation of the policy. Incidentally, it means we should use raw REST API for creating it, not a strongly typed ES client.
  2. yes
  3. My proposal is that we make strongly opinionated choice and only support one way of running ES with data streams and ILM. Deprecate all other modes.
  4. Another reason why I don't want to deal with that at all - if they want complex policy, they can create a complex JSON and either have Jaeger execute it on start-up or even execute it themselves.
  5. I don't know if data streams are compatible with any other ways. We recently landed a breaking change #6519 which has a migration path by manually creating (once) index aliases. We could investigate if similar steps are available for backwards compatibility, but if not we do it the new way. Most of the storage implementation doesn't actually change, so if people already have a legacy setup when it might still work.

**Comment by hartfordfive** [comment]
Hi all.  Any recent updates on this?   

Having the ability to set the backing Elastisearch indices as datastreams could also allow for significant reduction in storage with the recent [logsdb mode](https://www.elastic.co/search-labs/blog/elasticsearch-logsdb-index-mode) which was first released in v8.15 and is now stable as of [8.17](https://www.elastic.co/blog/whats-new-elastic-8-17-0).   This becomes especially valuable for deployments which receive a large number of traces.

**Comment by Manik2708** [comment]
> Hi all. Any recent updates on this?
> 
> Having the ability to set the backing Elastisearch indices as datastreams could also allow for significant reduction in storage with the recent [logsdb mode](https://www.elastic.co/search-labs/blog/elasticsearch-logsdb-index-mode) which was first released in v8.15 and is now stable as of [8.17](https://www.elastic.co/blog/whats-new-elastic-8-17-0). This becomes especially valuable for deployments which receive a large number of traces.

Currently Elastic search storage is going under heavy refactoring as part of #7034 to upgrade it to v2 APIs. Once it is completed, we can focus on data streams!

**Comment by malayh** [comment]
@Manik2708 Hi, Is there any update regarding this?

**Comment by Manik2708** [comment]
@malayh Currently I am very busy in my job hunt. Please go ahead if you can!

**Comment by anhtaw** [comment]
@Manik2708 First of all, thank you very much! I really appreciate the new Elasticsearch data stream feature — I hope it will be rolled out soon, as I really need it right now.

**Comment by Manik2708** [comment]
@yurishkuro I want to continue working on this issue. Last time where I left was when documentation was needed for my plan for implementation of ILM/ISM policy #6604 and comment https://github.com/jaegertracing/jaeger/pull/6604#issuecomment-2648072710. Now I saw that Jaeger want to move away from olivere client. So should I move from where I left? Also do we have some script for giving data to jaeger for testing? Or do we need to do that manually? Can we work over automating this (if not exist)?

**Comment by Manik2708** [comment]
@yurishkuro I have prepared a doc: https://docs.google.com/document/d/1WDQJmHGjnyck5h1DDvZf4ZJnTaXEF-wUvR6k_lipWJ4/edit?tab=t.0 for the proposal of data streams. Please have a look over this and provide your review (This is a design doc, once approved we can move to implementation part)

**Comment by TsengSR** [comment]
Has there been any progress recently? Datastreams would really be awesome as they make life so much easier. 

For the policies mentioned above: You don't really need any customization of the ILM Policy within jaeger itself. Either pick on of the (non-deprecated) Policies as default or create one that you think best suits the current behavior (i.e. daily roll-over, delete after x-day, move to warm/cold etc).

There needs be **no configuration or customization**  of the ILM Policy in jaeger itself. In ElasticSearc, users can customize streams and/or policies via templates, i.e. they create their own distinct policy the way they want it, then they create a `<datastreamname>@custom` template and set the new policy in there. 

This will automatically override given properties (i.e. ILM policy) of the underlying template and apply it. In ELK Stafck for example I use `kubernetes@custom` component template, to customize all ILM Policies of the Kubernetes Integration, and they apply my ILM Policy for it

```json
{
  "index": {
    "lifecycle": {
      "name": "14-days@atomicgear"
    },
    "default_pipeline": "drop_host_metadata"
  }
}
```

This `kubernetes@custom` Component Template (there are also Index Templetes for specifc indices if you dont want to apply a policy to multiple indices and a pipeline which removes some unnecessary host meta data). On next index creation, this will use the new template settings.

And the best of it, the user have no need to change the default jaeger policy ever and never risk that jaeger would override their settings in a future update. This also gives you the flexibility in future to adjust the default template and just updating the ILM policy when needed. It will not affect users customization as the `@custom` templates are applied after the default policy. 

This is ELKs recommended way to manage datastream customization. All thats needed would be a sample in the documentation, nothing else. 

See also [Customize built-in ILM policies in Elasticsearch](https://www.elastic.co/docs/manage-data/lifecycle/index-lifecycle-management/tutorial-customize-built-in-policies#example-using-index-lifecycle-policy-apply-policy)

**Comment by Manik2708** [comment]
@TsengSR Thanks for your suggestion! But I think you mixed the Datastream and ILM approaches. These two are different! In ILM approach we need to define rollover indices and some other parameters but the subtle difference in datastream is that it automates all the creation of rollover indices etc. Currently Jaeger supports the ILM way, you can turn on `UseILM` option to use the ILM policies but this issue aims to use datastreams which will give responsibility of creating policy and storing data to datatream to jaeger. We have some issues and backward compatibility concerns regarding making datastreams as default storage strategy, once that is solved, we will start implementing streams. Please see that this is what I could understand from your comment, are you suggesting something else regarding ILM? Would be happy to hear that!

**Comment by TsengSR** [comment]
@Manik2708 

> 2. Using ILM with data streams would be ideal but auto creating policy is challenging, as the policy for backing stream indexes and automatic rollover will look like (as per official docs):
> 
> ...
> 
> Now making the policy configurable for the user is going to be a challenge. We can choose to create our own policy struct (because it is not provided by ES-Client) but that will lead to maintainence overhead as keeping it always synchornized with ES updates will be challenging. 
> 3. What part of policy needs to be configurable? I am not in support of creating the whole policy configurable, take an example when using data streams we should not skip rollover. 

I specifically mean this comment (and other comments about the topic. 

One doesn't need to proviate **any** type of customization in jaeger. Nothing. Just none. 

All that jaeger would have to do is to define one standard policy it wants to use (which can be one of the out-of-the-box policies from Elasticsearch or one you define). 

It can be static and must only be rolled out once when deployed (and whenever it changes). It (+ the templates) can be pushed on every start of jaeger collector overriding the existing ones without any issues. 

Assume, the ILM policy is called `jaeger-ilm-policy` and the stream is called `jaeger-span` and `jaeger-service` (named after the current index names). 

Now if you want to customize `jaeger-span` the user goes to Kibana and creates a new policy (or uses the Dev Tools to do via rest).

Then the user creates a new component template, like `jaeger-span@custom` via Kibana > Management > DevTools and runs 

```
PUT _component_template/jaeger-span@custom
{
  "template": {
    "settings": {
      "index": {
        "lifecycle": {
          "name": "14-days-rollover"
        }
      }
    },
    "mappings": {
      "properties": {
      }
    }
  }
}
```

From now on, next time the index is rolled over, the `14-days-rollover` ILM policy will be assigned to the ~~datastream~~ newly created index. 

The advantages of this, there is no need to edit the `jaeger-ilm-policy` provided by jaeger nor custome it, neither to edit the default template with the risk of it being overwritten on future update. 

When a future update happens, jaeger team can simply change the default template and ilm policy as they seem fit, without effecting people using their `@custom` policies. The `<datastream>@custom` or `<integration>@custom` is a naming convention. 

When ELK sees, there is a `jaeger-span@custom` template, it will appy it to `jaeger-span` datastream automatically. 

This gives users full control and the application developers an option to only maintain default/recommended ilm/templates and not to worry how the user can customize it because a mechanism is already provided by elasticsearch itself. 

TL;DR: Customization of templates and definition if ILM is done solely in Elasticsearch/ELK via name-convetion templates and is not a concern of the applications writing to it. 

The only thing needed about customization is a mention in the documentation with an example how to create an template and/or your own ILM and how to assign it using the `@custom` name convention.

**Comment by Manik2708** [comment]
@yurishkuro Can we look over this approach?

**Comment by yurishkuro** [comment]
@Manik2708 sounds great if it works like that.

**Comment by Manik2708** [comment]
@TsengSR Thanks a lot for this explanation, what I could understand from your comment is that users must be given proper steps in documentation for changing the default templates and ilm policy or more specifically it aims to solve the issue of customization of ILM policy. Am I right?

**Comment by hartfordfive** [comment]
One more important aspect to note regarding elasticsearch datastreams is the ability to enable [logsdb](https://www.elastic.co/search-labs/blog/elasticsearch-logsdb-index-mode) mode if using version 8.17 or higher.  I've experienced over a 50% reduction is size on disk using this feature with logs so this definitely is a big added benefit.  The searches on datastreams with logsdb may also result in lower query latencies due to the sorting data in the indices (at indec time) by the "@timestamp" and "host.name" fields, but the latter can be changed to a field which is present in the span index mapping.  There can be a slight increase in CPU usage when enabling logsdb mode but in my experience it's been minimal and more than worth the benefit.

**Comment by TsengSR** [comment]
> [@TsengSR](https://github.com/TsengSR) Thanks a lot for this explanation, what I could understand from your comment is that users must be given proper steps in documentation for changing the default templates and ilm policy or more specifically it aims to solve the issue of customization of ILM policy. Am I right?

@Manik2708 
Basically yes. The ElasticSearch/ELK Documentation is extensive, but its really not easy to find stuff in there. After I tried to edit the indices/streams directly and got everything overwritten on an ELK/Integrations Update, I spent like 2 days trying to find a more permanent solution and came across the `@custom` name convention. Worked since then. 

Basically one doesn't edit the defaults, one defines a new one with `@custom` which overrides the properties defined in there. They basically get combined (default one + `@custom` one, where `@custom` takes precedence.

It can be override by stream, index and per integration (or namespace, dunno the term). In Kubernetes case a `kubernetes@custom` will be applied to all streams that come from the kubernetes integration like `kubernetes.apiserver-*`, `kubernetes.volume*` etc.

**Comment by SoumyaRaikwar** [comment]
@yurishkuro  

I've updated the design document and the [Gist](https://gist.github.com/SoumyaRaikwar/33b696ae9ce488c99c165f6301adf7e1) with the proposed **Index Lifecycle Management (ILM)** policy for data streams.

**Updates include:**
-   **ILM Policy Proposal (`jaeger-ilm-policy`)**:
    -   **Hot Phase**: Rollover at 50GB or 1 day. Priority 100.
    -   **Warm Phase**: Transition immediately after rollover. Priority 50.
    -   **Delete Phase**: Delete indices after 7 days.
-   **Technical Verification**:
    -   Updated [jaeger-span-8.json](cci:7://file:///media/soumya/DATA/open-source/jaeger/internal/storage/v1/elasticsearch/mappings/jaeger-span-8.json:0:0-0:0) template to reference the ILM policy.
    -   Verified that all mapping tests pass with the new configuration.
    -   Created a sample [jaeger-ilm-policy.json](cci:7://file:///media/soumya/DATA/open-source/jaeger/jaeger-ilm-policy.json:0:0-0:0) definition.

Please review the updated proposal. Once approved, I can proceed with applying the ILM policy to the templates.

**Comment by hartfordfive** [comment]
I would also suggest the the following minor changes to the ilm policy: 
- addition of  the forcemerge action in the [ILM warm phase](https://www.elastic.co/docs/manage-data/lifecycle/index-lifecycle-management/index-lifecycle#warm-phase) as this will optimize searches and reduce index overhead.   Force merging is a resource intensive process so it's best to have a dedicated warm node pool in which these operations will run so that they don't take away resources from the hot nodes which are actively indexing data. 
- additional to the 50gb size limit, also set a max doc limit of 200 million in the rollover action (max_docs: 200000000)

**Comment by SoumyaRaikwar** [comment]
@hartfordfive Thank you for the ILM policy suggestions! I've implemented both improvements:

##  Enhanced ILM Policy

### Hot Phase - Added max_docs limit
```json
"rollover": {
  "max_age": "1d",
  "max_primary_shard_size": "50gb",
  "max_docs": 200000000
}
```
This prevents indices from growing too large in document count, which is important for high-throughput tracing workloads.

### Warm Phase - Added forcemerge
```json
"forcemerge": {
  "max_num_segments": 1
}
```
This optimizes search performance by reducing segment overhead. Since warm indices are read-only after rollover, forcemerging to a single segment provides better query performance without impacting indexing.

changes are now in:
- Updated [Gist](https://gist.github.com/SoumyaRaikwar/33b696ae9ce488c99c165f6301adf7e1)

The combination of these settings ensures optimal resource usage on both hot and warm nodes. Let me know if you'd like any adjustments to the thresholds!

**Comment by hartfordfive** [comment]
Is it necessary to rollover at a max age of 1d?   If having daily indices isn't necessary for jaeger, I would also suggest simply removing that `max_age` directive.   In cases where users are potentially running smaller setups, there is a risk they don't accumulate 50GB of data in the index within the same day.   You might also have users which have considerable traces during weekdays and minimal volume, if any during weekends.   In my experience, it's best to just let the shards grow up to 50gb or 200 million documents unless there's a clear justification for adding in an age limit.   

Another thing to note is that some users may be using the licensed version of Elasticsearch.   In that case, they have access to [searchable snapshots](https://www.elastic.co/docs/deploy-manage/tools/snapshot-and-restore/searchable-snapshots), which further decreases disk usage by cutting out the need for a replica shard once the index has been persisted in a snapshot.  Might be beneficial to also include an alternate ILM policy for licensed Elasticsearch installations which simply adds the `searchable_snapshot` action in the hot, cold or frozen phase (see [docs](https://www.elastic.co/docs/reference/elasticsearch/index-lifecycle-actions/ilm-searchable-snapshot)).   For example, it could look something like this:

```json
{
  "policy": {
    "phases": {
      "hot": {
        "min_age": "0ms",
        "actions": {
          "rollover": {
            "max_primary_shard_size": "50gb",
            "max_docs": 200000000
          },
          "set_priority": {
            "priority": 100
          }
        }
      },
      "cold": {
        "min_age": "0ms",
        "actions": {
          "searchable_snapshot" : {
            "snapshot_repository": "my_ilm_repo"
          },
          "set_priority": {
            "priority": 50
          }
        }
      }
      "delete": {
        "min_age": "7d",
        "actions": {
          "delete": {
            "delete_searchable_snapshot": true
          }
        }
      }
    }
  }
}

```

Take note that the `forcemerge` action has been removed from the `warm` phase as it's automatically performed as part of the searchable_snapshot phase by default.  Take note that if a user has a high volume of searches on the traces, then they may opt to set the `replicate_for` config in the searchable_snapshot phase to indicate the duration which a replica shard will be kept before being reduced back down to 0 replicas (default).

It would be great if Jaeger also documented the option of using an OTEL collector instance to add the `@timestamp` field on all documents prior to indexing to Elasticsearch.    The reason I bring this up is again relating to users with the Elastic enterprise license, specifically those orchestrating with [ECK](https://www.elastic.co/docs/deploy-manage/deploy/cloud-on-k8s).     In this case, regardless of the Elasticsearch [node type](https://www.elastic.co/docs/deploy-manage/distributed-architecture/clusters-nodes-shards/node-roles#transform-node-role) (ingest, master, data, ml, etc.), they all count against the overall license usage so while ingest nodes do add some convenience, it often comes at a cost which I don't find justifiable in my experience.   If using Kafka to initially buffer the traces, you can avoid using Elasticsearch ingest nodes by simply consuming the spans form kafka and then apply transformation operations to add the required fields just prior to sending them to the [elasticsearch exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/elasticsearchexporter).

One more thing I've realized is that I don't see a proposition to also enable the logsdb mode in the existing mappings for ES versions 8.17 or higher (for example, [jaeger-span-8](https://github.com/jaegertracing/jaeger/blob/main/internal/storage/v1/elasticsearch/mappings/jaeger-span-8.json)).   To take advantage of this, you'll need to enable the logsdb mode:
```
  "template": {
     "settings": {
        "index.mode": "logsdb"
     }
  },
```
although you'll need to encapsulate that in a conditional statement so that users can explicitly enable it if they're running version 8.17 or higher.  In the meantime, I invite the core Jaeger maintainers to test out the logsdb mode as I would be curious to know what kind of space reductions are observe in a test environment.  


My apologies for all this back and forth.  I just figured I'd provide some context based on experience as I suspect it could be useful for others!   If any other experienced Elasticsearch users have other opinions and/or feedback, I would be more than glad to hear those also.

**Comment by SoumyaRaikwar** [comment]
@hartfordfive,
 I've incorporated all of them into the design and implementation:

1.  **ILM Optimizations**:
    -   **Removed `max_age: 1d`**: I agree that letting shards grow to optimal size (50GB/200M docs) is better for most Jaeger installations, especially smaller ones. I've updated the default policy accordingly.
    -   **Licensed ILM Policy**: I've added an "Alternate (Licensed/Enterprise)" ILM policy section to the design document that includes the `cold` phase with `searchable_snapshot` and removes unnecessary `forcemerge` (as it's part of the snapshot phase).

2.  **OTEL Collector Pre-processing**:
    -   I've added a section explaining how to use the `transform` processor in an OTEL collector to add the `@timestamp` field before indexing. This is a great alternative for users looking to avoid ingest node license costs in ECK.

3.  **LogsDB Mode (ES 8.17+)**:
    -   I've implemented a new configuration option `enabled_logsdb`. When enabled, the ES 8 templates will now conditionally include `"index.mode": "logsdb"`.
    -   I've also added unit tests to ensure this renders correctly for ES 8+ templates while remaining optional as suggested.

I've updated the [design doc](https://gist.github.com/SoumyaRaikwar/33b696ae9ce488c99c165f6301adf7e1) with these changes for your review and approval. Looking forward to your thoughts!

**Comment by hartfordfive** [comment]
I realize this could take time for these changes are applied although do we have an estimate as to when it might be available?   I'm looking forward to reducing the overall storage size of the span indices.

**Comment by TsengSR** [comment]
@SoumyaRaikwar 

> [@yurishkuro](https://github.com/yurishkuro)
> 
> I've updated the design document and the [Gist](https://gist.github.com/SoumyaRaikwar/33b696ae9ce488c99c165f6301adf7e1) with the proposed **Index Lifecycle Management (ILM)** policy for data streams.

May be worth considering to have a dot-notion in the stream names, 

Instad of 

> `jaeger-ds-span`: Spans data stream.
> `jaeger-ds-service`: Service/Operation data stream.
> `jaeger-ds-sampling`: Sampling probabilities data stream.
> `jaeger-ds-dependencies`: Dependency data stream.

have


> `jaeger.span`: Spans data stream.
> `jaeger.service`: Service/Operation data stream.
> `jaeger.sampling`: Sampling probabilities data stream.
> `jaeger.dependencies`: Dependency data stream.

From what I see, this would allow to define a template that overrides all 4 of th em, i.e. using my example from above to set ILM policy

```
PUT _component_template/jaeger@custom
{
  "template": {
    "settings": {
      "index": {
        "lifecycle": {
          "name": "14-days-rollover"
        }
      }
    },
    "mappings": {
      "properties": {
      }
    }
  }
}
```

This would set the `14-days-rollover` pollicy on all 4 datastreams, which makes managing of data streams easier Of one wants set one of them differently, one uses `jaeger.span@custom` (`<datastream>@custom`) as tempalte name.

**Comment by tronda** [comment]
I guess that this feature is focused on ElasticSearch, but I wonder if most of these concepts also apply to OpenSearch and its [ISM policies](https://docs.opensearch.org/latest/im-plugin/ism/index/) and [DataStream support](https://docs.opensearch.org/latest/im-plugin/data-streams/). OpenSearch also support [searchable snapshots](https://docs.opensearch.org/latest/tuning-your-cluster/availability-and-recovery/snapshots/searchable_snapshot/) without requiring any license. OpenSearch has [Application-Based Configuration](https://opensearch.org/blog/opensearch-simplified-the-power-of-application-based-templates/) where one can have [index templates which is optimized for the log use case](https://docs.opensearch.org/latest/im-plugin/index-context/#create-an-index-template). Any thoughts on how much of the code would be able to utilize these features in OpenSearch?

**Comment by yurishkuro** [comment]
The solution definitely needs to work for both ES and OS.

**Comment by Manik2708** [comment]
@hartfordfive There could be many more customization to ILM policy, what I think is that Jaeger should provide only a basic ILM policy by default and rest all configurations should be left to the user as managing ILM policies is beyond the scope of Jaeger. Customization plan of @TsengSR looks good, I think we can go with that! @SoumyaRaikwar are you still working on #7768?

**Comment by Manik2708** [comment]
> I guess that this feature is focused on ElasticSearch, but I wonder if most of these concepts also apply to OpenSearch and its [ISM policies](https://docs.opensearch.org/latest/im-plugin/ism/index/) and [DataStream support](https://docs.opensearch.org/latest/im-plugin/data-streams/). OpenSearch also support [searchable snapshots](https://docs.opensearch.org/latest/tuning-your-cluster/availability-and-recovery/snapshots/searchable_snapshot/) without requiring any license. OpenSearch has [Application-Based Configuration](https://opensearch.org/blog/opensearch-simplified-the-power-of-application-based-templates/) where one can have [index templates which is optimized for the log use case](https://docs.opensearch.org/latest/im-plugin/index-context/#create-an-index-template). Any thoughts on how much of the code would be able to utilize these features in OpenSearch?

For Open Search the APIs remain same but they differ in ISM policy and templates. We could maintain a separate ISM policy and mutate the template to make it compatible for OpenSearch with ISM policies. Optimizing for logs could be a different issue and can be resolved after addition of datastream. The snapshot feature is an index template customization, we can discuss that whether it should be from user side or could be added to Jaeger config.

**Comment by SoumyaRaikwar** [comment]
> [@hartfordfive](https://github.com/hartfordfive) There could be many more customization to ILM policy, what I think is that Jaeger should provide only a basic ILM policy by default and rest all configurations should be left to the user as managing ILM policies is beyond the scope of Jaeger. Customization plan of [@TsengSR](https://github.com/TsengSR) looks good, I think we can go with that! [@SoumyaRaikwar](https://github.com/SoumyaRaikwar) are you still working on [#7768](https://github.com/jaegertracing/jaeger/pull/7768)?

@Manik2708 yes i am working on that and will update the [design doc](https://gist.github.com/SoumyaRaikwar/33b696ae9ce488c99c165f6301adf7e1) as per feedback.

**Comment by SoumyaRaikwar** [comment]
@yurishkuro Thanks for the feedback.

I've updated the [design document](https://gist.github.com/SoumyaRaikwar/33b696ae9ce488c99c165f6301adf7e1) to address the requirements for both Elasticsearch and OpenSearch support in a unified codebase.

**Implementation Strategy:**

1.  **Uniform Naming**:
    *   Switching to dot-notation (e.g., `jaeger.span`, `jaeger.service`) for both ES and OS. This avoids conflicts with legacy patterns and aligns with ecosystem standards.

2.  **Shared vs. Divergent Paths**:
    *   **Writing Data**: The ingestion logic remains identical for both backends. The writer simply sends documents to the data stream alias (`jaeger.span`).
    *   **Lifecycle Management (Divergent)**:
        *   **Elasticsearch**: Uses **ILM** policies.
        *   **OpenSearch**: Uses **ISM** policies.
    *   **Templates**: The template creation logic will dynamically inject the correct policy definition (ILM or ISM) based on the configuration.

3.  **Configuration**:
    We are introducing flags to explicitly control the behavior:
    *   `UseDataStream`: Enables the data stream mode (dot-notation, alias writing).
    *   `UseISM`: Explicitly enables OpenSearch ISM support.
    *   `UseILM`: Explicitly enables Elasticsearch ILM support.

    *Example*:
    *   `UseDataStream=true` + `UseILM=true` → Standard ES Data Streams.
    *   `UseDataStream=true` + `UseISM=true` → OpenSearch Data Streams with ISM.

I will proceed with the implementation based on this design once approved.

**Comment by SoumyaRaikwar** [comment]
@yurishkuro  should i start  implementing according to [design document](https://gist.github.com/SoumyaRaikwar/33b696ae9ce488c99c165f6301adf7e1) or there's something you would like me to change

**Comment by repairman29** [comment]
🎯 **PING RECEIVED!**

**@repairman29** has claimed this bounty via [Echeo](https://echeo.io).

**Match Score:** 85%  
**Status:** Ready to start working!

[View conversation in Echeo](https://echeo.io/messages) to coordinate.

**Comment by repairman29** [comment]
Hi! I'm interested in helping with this issue. 

I have experience with index (repairman29/ai-gm-service) and think I can contribute here. Would love to discuss the approach and see if we can make this happen.

*(Connected via [Echeo](https://echeo.io) - a platform for matching developers with open source bounties)*

**Comment by Manik2708** [comment]
@yurishkuro My steps for this issue goes like:
1. Deprecating `CreateIndexTemplates` and set it to `true` by default. This will be followed by providing a doc to users who are using their custom templates to use `.custom` strategy to update the default mapping. The reason for this is because datastream issue will require some template changes and I suggest that we should move to this strategy because in future if we again need to change the template it will be effortless.
2. Make Jaeger ready to automatically create an ILM/ISM policy on start.
3. Enable Datastream for ES version 8 (I am updating the docs about ingestion pipelines, I don't think we require them).
4. Enabling LogsDb and other features for Datastream.

Open for questions over this flow!

**Comment by yurishkuro** [comment]
@Manik2708 & @SoumyaRaikwar - what is the authoritative design doc? I see two, an .md file and a google doc. Let's have a single google doc for now, and once we iron out the open questions we can convert it into an ADR.

**Comment by Manik2708** [comment]
@SoumyaRaikwar Please add your proposal to the https://docs.google.com/document/d/1WDQJmHGjnyck5h1DDvZf4ZJnTaXEF-wUvR6k_lipWJ4/edit?tab=t.0

**Comment by SoumyaRaikwar** [comment]
@yurishkuro ,@Manik2708 
i have added the proposal to https://docs.google.com/document/d/1WDQJmHGjnyck5h1DDvZf4ZJnTaXEF-wUvR6k_lipWJ4/edit?tab=t.0

**Comment by SoumyaRaikwar** [comment]
@yurishkuro, @jkowall

I've updated the proposal https://docs.google.com/document/d/1WDQJmHGjnyck5h1DDvZf4ZJnTaXEF-wUvR6k_lipWJ4/edit?tab=t.0 with the following additions:

- Executive Summary - Added Hybrid Model recommendation upfront (Spans → Data Streams, Services/Dependencies → Standard Indices)
- OpenSearch Compatibility - ISM/ILM is automatically detected based on backend (no additional flag needed)
- Scope Clarification - Explicitly stated Services & Dependencies stay as indices, eliminating @timestamp concerns
- New Configuration Flag - --es.use-data-stream to enable Data Streams for Spans
- Implementation Phases - Added phased rollout plan (Core → Migration → Stabilization)

These additions reconcile the new design decisions with the existing detailed proposal. Please review and let me know if any changes are needed

**Comment by Manik2708** [comment]
@yurishkuro Should I raise the PR for ADR so that AI could review it?

**Comment by yurishkuro** [comment]
no, I can have Claude review it locally, let me do that.

**Comment by yurishkuro** [comment]
# Critique: DataStream Proposal — Backwards Compatibility Gaps

## Gaps in Operational Mode Coverage

The proposal defines **4 operational modes** in "Current Behaviour" but the "User Story" and "Jaeger Story" sections don't cleanly cover all of them.

**Mode 3 (UseReadWriteAliases=true, UseILM=false — manual rollover via cron/`es-rollover`)** is absent from the User Story section. It is mentioned obliquely in "Jaeger Story" point 1, but never explicitly addressed as a migration path. This is a real production use case.

---

## Specific Backwards Compatibility Gaps

### 1. The `jaeger-span-all-read` alias construction is buggy

The proposed alias:
```json
{ "add": { "index": "jaeger-span-*", "alias": "jaeger-span-all-read" }},
{ "add": { "index": "jaeger-span-ds", "alias": "jaeger-span-all-read" }}
```
The wildcard `jaeger-span-*` will match `jaeger-span-ds` (the new data stream), resulting in a duplicate entry. More importantly, this alias creation at Jaeger startup is not idempotent if old indices do not exist yet (e.g., new deployments). The proposal does not address who creates this alias, when, how failures are handled, or what happens on a restart.

### 2. `jaeger-es-index-cleaner` tool — completely unaddressed

Mode 1 users rely on `jaeger-es-index-cleaner` for TTL-based deletion of daily indices. Data streams do not support deleting individual backing indices directly — deletion goes through ILM or the data stream delete API. The proposal does not mention whether this tool will be updated, deprecated, or will silently break for users who depend on it.

### 3. `es-rollover` binary fate is unclear

The `es-rollover init` job creates aliases and applies ILM policies for Mode 3 and Mode 4 users. Under the new model this job is either unnecessary or needs to change its behavior. Users who have operational scripts or Helm jobs calling `es-rollover` need explicit migration guidance. The proposal says to deprecate `UseILM` but does not say what replaces the rollover binary's responsibilities.

### 4. No rollback strategy

If a user upgrades Jaeger to the data-stream version and encounters a problem, the previous Jaeger version will not know about `jaeger-ds-span` and cannot read from it. The proposal has no mention of a downgrade path or what ES state is left behind if a rollback is needed.

### 5. Zero-downtime rolling upgrade gap

During a rolling upgrade (old pod and new pod running simultaneously), the old pod writes to `jaeger-span-YYYY-MM-DD` and the new pod writes to `jaeger-ds-span`. Reads during this window are inconsistent unless `DualLookUp` is already enabled before the first new pod starts. The required sequencing and any locking or coordination between versions is not described.

### 6. DualLookUp disablement assumes homogeneous retention

Phase 4 says: *"After retention period (e.g., 14 days), old indices age out."* This assumes all users have a 14-day retention policy. Many users retain data for 30, 90, or 180+ days. Disabling DualLookUp on a fixed schedule would silently drop visibility into old data for those users. The disablement should be keyed off actual index existence, or at minimum the user's configured retention period — not a hardcoded time assumption.

### 7. Index prefix / multi-tenancy not addressed

The templates correctly use `{{ .IndexPrefix }}`, but the new alias `jaeger-span-all-read` is shown without a prefix variable. Users running multiple Jaeger instances against a shared ES cluster — each with a different index prefix (e.g., `prod-`, `staging-`) — would have alias collisions at startup. This is a significant gap for any non-trivial deployment and needs to be addressed explicitly.

### 8. Downstream consumers (Kibana index patterns, Grafana datasources)

Many users have Kibana index patterns or Grafana Elasticsearch datasources pointing at `jaeger-span-*`. Changing the primary write target to `jaeger-ds-span` without addressing this means their dashboards and visualizations stop showing new data immediately after upgrade. The proposal is silent on this class of user.

### 9. Services/Dependencies alias path under Mode 3/4

The proposal states Services and Dependencies remain as standard indices. However, for Mode 3 and Mode 4 users, `jaeger-service-read` and `jaeger-service-write` aliases exist and are actively used. The proposal does not confirm whether these aliases are preserved as-is, or clarify how the "no changes to services/dependencies" decision interacts with the existing alias setup.

### 10. `CreateIndexTemplates` deprecation creates a footgun

User Story 3 proposes deprecating `CreateIndexTemplates` (defaulting it to `true`). Users who set it to `false` did so intentionally — commonly because their templates are managed externally via Terraform, Helm, or the Elastic operator. Silently enabling template creation on upgrade would overwrite or conflict with those externally managed templates. This needs either an explicit conflict-detection mechanism (detect existing template, refuse to overwrite) or a hard startup warning that requires opt-in acknowledgment.

---

## Structural Issue

The "Jaeger Story" section point 2.3 states: *"The story of new users would remain the same as old users not using ILM."* This is backwards — new users should be the baseline, not derived from legacy behavior. It also implies new users receive the old-style template, which contradicts Phase 3 of the implementation plan (new `jaeger-ds-span` template). This section needs a rewrite for clarity.

---

## Summary

| Area | Status |
|---|---|
| Mode 3 (manual rollover, no ILM) | Not explicitly covered |
| `jaeger-es-index-cleaner` tool impact | Unaddressed |
| `es-rollover` binary migration | Unclear |
| Rollback / downgrade path | Missing |
| Zero-downtime rolling upgrade | Missing |
| DualLookUp disablement tied to retention | Assumes 14d for all users |
| Index prefix / multi-tenancy | Partial (templates only, alias hardcoded) |
| Downstream Kibana/Grafana consumers | Missing |
| Services/Dependencies alias path (Mode 3/4) | Ambiguous |
| `CreateIndexTemplates` conflict detection | Missing |

**Comment by Manik2708** [comment]
1. There was a typo of `jaeger-span-ds`, it is decided in 4th step that new index pattern will be `jaeger-ds-span`. Also this alias approach is removed as for now because queries required for datastream data and old data are different.
2. Datastreams aim to deprecate every es tool that includes cleaner too. Deletion will be handled completely by ILM policy.
3. Yes this tool will be deprecated and the deprecation PR will consist of every migration step required for the users. This tool will become unnecessary after datastreams, so we don't need to think about its responsibilities. 
4. I haven't much to say about this, the best solution is creation of a script for downgrade which will transfer all the datastream data back to legacy data pattern so that user can rollback. Maybe @yurishkuro can help with this.
5. Can't understand the question, the feature gate will be enabled by default so the new pod will always read from both old and new data that means even if old pod is writing to old data, the new pod will be reading that too.
6. Agreed, therefore removed it. The dual-lookup is independent of retention period, it will be disabled in 4 release cycles.
7. Yes! Therefore this alias approach is removed as a whole and queries will themselves read from old and new data.
8. @yurishkuro I am unaware of this case, do you have any thoughts over this? We can provide a proper documentation regarding the new change so that user are well aware of this so as to change the sources and make rest of the things compatible with datastreams.
9. In long term we want to eliminate the rollover and other es cmds but we can't enable datastreams for services/dependencies because it is not time-series data but we will automate the rollover with ILM approach. Therefore it is not a part of datastream issue but a separate one.
10. The reason we want to get rid of this parameter is because it makes very hard for developers to do any change to template. Take an example of introducing scope parameters is blocked because it will become a breaking change and reason is that we need to give migration steps every time a change is introduced. Users could still be able to change the templates but through component templates. So in the deprecation PR (which will be the first PR for this issue), a migration guide will be given to all the users (this is mentioned in proposal) to make a backup of old templates and change the templates using component templates.

**Comment by yurishkuro** [comment]
Status note: the `rotation.data_stream` activation work (RFC 0004 Phase 2) is now **unblocked**.

It had been paused pending RFC 0006 (unified/owned Elasticsearch client), because the template/policy bootstrap would otherwise have been built on the soon-to-be-removed app-layer `mappings` renderer. The relevant milestone — RFC 0006 **M4b**, `esclient` owns index templates (#8963) — has now merged, so that foundation is settled.

Where things stand:
- #8833 — DataStreamRotation strategy + data-stream-aware write path — **merged, but dormant** (config still rejects `data_stream` as "not yet implemented").
- #8873 — ctx/request refactor (first carved-out slice) — **merged**.
- #8880 — `@settings` component template (second slice) — **closed**, superseded by M4b.

Remaining to make the feature functional: rebuild the composable `_index_template` + `@mappings`/`@settings` component templates on the owned esclient, add the default ISM/ILM policy, flavor detection, factory startup wiring, flip the validation gate for the spans index, and add the integration test. Ready to resume as a small stack on top of the new client.
